client/src/components/StreamField/blocks/StructBlock.js

```diff
@@ -96,9 +101,28 @@ export class StructBlock {
     }
     const error = errorList[0];
 
-    for (const blockName in error.blockErrors) {
-      if (hasOwn(error.blockErrors, blockName)) {
-        this.childBlocks[blockName].setError(error.blockErrors[blockName]);
+    // Non block errors
+    const container = this.container[0];
+    container
+      .querySelectorAll(':scope > .help-block.help-critical')
+      .forEach((element) => element.remove());
+
+    if (error.nonBlockErrors) {
```

Comment:
Arrays always evaluate to `true` in JavaScript, even if it's empty. Even though the only code inside this block is a `.forEach()` statement which will be skipped if the array is empty, let's explicitly check for the length:

```suggestion
    if (error.nonBlockErrors && error.nonBlockErrors.length > 0) {
```

If we're certain that `error.nonBlockErrors` is an array (not `null`/`undefined`), then  we can just check the `.length` directly, like how we do it for `StreamBlock`:

https://github.com/wagtail/wagtail/blob/dc2afe89b94265ea1dd592ea0e64090b05de2cf8/client/src/components/StreamField/blocks/StreamBlock.js#L418

Reply:
This logic has evolved in a fairly ad-hoc way over time, and so I'm wary of making any assumptions about whether the `nonBlockErrors` key is present or not - the `if (error.nonBlockErrors)` here is indeed to guard against the possibility if the key being absent.

In the refactored version in #10131, the `messages` key (which takes over the job of `nonBlockErrors`) is explicitly allowed to be omitted, which means there shouldn't ever be a reason for an empty list to be passed here - and, as you note, if that does happen then the `.forEach` will do the right thing anyhow - so at least in the post-#10131 code, I think it does make sense to omit the length check.

Reply:
Ah, that makes sense!